tezos/new_context/src/timings.rs

```diff
@@ -49,10 +69,545 @@ pub fn context_action(
+        current: &mut Option<(HashId, T)>,
+        table_name: &str,
+    ) -> Result<(), SQLError>
+    where
+        T: Eq,
+        T: AsRef<Vec<u8>>,
+    {
+        match (hash.as_ref(), current.as_ref()) {
+            (None, _) => {
+                *current = None;
+                return Ok(());
+            }
+            (Some(hash), Some((_, current_hash))) if hash == current_hash => {
+                return Ok(());
+            }
+            _ => {}
+        };
+
+        let hash = hash.unwrap();
+        let hash_string = Self::hash_to_string(hash.as_ref());
+
+        if let Some(id) = Self::get_id_on_table(sql, table_name, &hash_string)? {
+            current.replace((id.to_string(), hash));
+            return Ok(());
+        };
+
+        sql.execute(
+            &format!(
+                "INSERT INTO {table} (hash) VALUES (?1);",
+                table = table_name
+            ),
+            [&hash_string],
+        )?;
+
+        let id = Self::get_id_on_table(sql, table_name, &hash_string)?
+            .expect("Unable to find row after INSERT"); // This should never happen
+
+        current.replace((id.to_string(), hash));
+
+        Ok(())
+    }
+
+    fn get_id_on_table(
+        sql: &Connection,
+        table_name: &str,
+        hash_string: &str,
+    ) -> Result<Option<i64>, SQLError> {
+        let mut stmt = sql.prepare(&format!(
+            "SELECT id FROM {table} WHERE hash = ?1;",
+            table = table_name
+        ))?;
+        let mut rows = stmt.query([hash_string])?;
+
+        if let Some(row) = rows.next()? {
+            return Ok(Some(row.get(0)?));
+        };
+
+        Ok(None)
+    }
+
+    fn insert_checkout(
+        &mut self,
+        context_hash: ContextHash,
+        irmin_time: f64,
+        tezedge_time: f64,
+    ) -> Result<(), SQLError> {
+        self.checkouts.add(irmin_time, tezedge_time);
+        self.set_current_context(context_hash)?;
+        self.insert_action(&Action {
+            name: "checkout".to_string(),
+            key: vec![],
+            irmin_time,
+            tezedge_time,
+        })
+    }
+
+    fn insert_commit(&mut self, irmin_time: f64, tezedge_time: f64) -> Result<(), SQLError> {
+        self.commits.add(irmin_time, tezedge_time);
+        self.insert_action(&Action {
+            name: "commit".to_string(),
+            key: vec![],
+            irmin_time,
+            tezedge_time,
+        })?;
+        self.update_global_stats()
+    }
+
+    fn insert_action(&mut self, action: &Action) -> Result<(), SQLError> {
+        let block_id = self.current_block.as_ref().map(|(id, _)| id.as_str());
+        let operation_id = self.current_operation.as_ref().map(|(id, _)| id.as_str());
+        let context_id = self.current_context.as_ref().map(|(id, _)| id.as_str());
+
+        let key = if action.key.is_empty() {
+            None
+        } else {
+            Some(action.key.join("/"))
+        };
+
+        self.sql.execute(
```

Comment:
Possible improvement here now that we use param placeholders instead of building the string by hand (because it is called very often, and we want it to be as fast as possible) is to use [prepare_cached](https://docs.rs/rusqlite/0.25.1/rusqlite/struct.Connection.html#method.prepare_cached). 

This will prepare the query the first time and then reuse the cached version on every subsequent call, avoiding the overhead of parsing the SQL query.